src/Npgsql/ConnectorPool.cs

```diff
@@ -413,86 +385,46 @@ internal void Release(NpgsqlConnector connector)
-                // count, while atomically make sure the waiting count isn't increased.
-                // Note that we also must update the state *before* putting the connector back in the idle list.
-                var newState = state;
-                newState.Idle++;
-                newState.Busy--;
-                CheckInvariants(newState);
-                if (Interlocked.CompareExchange(ref State.All, newState.All, state.All) != state.All)
-                {
-                    // Our attempt to decrement the busy count failed, either because a waiting attempt has been added
-                    // or busy has changed. Loop again and retry.
-                    continue;
+                    // If the open attempt timed out, the Task's state will be set to Canceled and our
+                    // TrySetResult fails. Try again.
+                    Debug.Assert(tcs.Task.IsCanceled);
+                    break;
                 }
+            }
 
-                // If we're here, we successfully applied the new state above and can put the connector back in the idle
-                // list (there were no pending open attempts).
-
-                connector.ReleaseTimestamp = DateTime.UtcNow;
-
-                // We start scanning for an empty slot in "random" places in the array, to avoid
-                // too much interlocked operations "contention" at the beginning.
-                var start = Thread.CurrentThread.ManagedThreadId % _max;
-
-                sw = new SpinWait();
-                while (true)
-                {
-                    for (var i = start; i < _idle.Length; i++)
-                    {
-                        if (Interlocked.CompareExchange(ref _idle[i], connector, null) == null)
-                        {
-                            Counters.NumberOfFreeConnections.Increment();
-                            return;
-                        }
-                    }
-
-                    for (var i = 0; i < start; i++)
-                    {
-                        if (Interlocked.CompareExchange(ref _idle[i], connector, null) == null)
-                        {
-                            Counters.NumberOfFreeConnections.Increment();
-                            return;
-                        }
-                    }
-                    sw.SpinOnce();
-                }
+            // Scenario: pre-empted release
+            // Right here between our check for waiters and our signalling decrement for storing
+            // a connector there could have been an awaiter enqueueing, we compensate at the end.
+
+            // If we're here, we put the connector back in the idle list
+            // We decrement Busy, any allocate that is racing us will not match Busy == _max
+            // and will not enqueue but try to get our connector.
+            Interlocked.Decrement(ref State.Busy);
+            connector.ReleaseTimestamp = DateTime.UtcNow;
```

Comment:
We could also squeeze L419-L422 into here, right after the decrement, but instead of handing it null actually hand it the connector? Need to compensate busy decrement with an increment again but it could be better?